File: integrations/mem0/typescript/src/workflow.ts
import { createHash } from "node:crypto";
import { Mem0Memory, memoryCaptureIntent, requiredText, positiveInteger } from "./memory.js";
import { MemoryError } from "./journal.js";
import type { MemoryCaptureAuthorization, MemoryOperation, MemoryRecord, MemoryResolution, MemoryReview } from "./memory.js";

export type MemoryDecisionPolicy = (candidate: MemoryRecord, review: MemoryReview) => Promise<MemoryResolution | undefined> | MemoryResolution | undefined;
export interface MemoryWorkflowResult {
  readonly extraction: MemoryOperation;
  readonly resolutions: readonly MemoryOperation[];
  readonly pendingIds: readonly string[];
}
function digest(value: string): string { return createHash("sha256").update(value).digest("hex"); }

/** Reuse the journal on retry; authorization remains an explicit host policy. */
export class MemoryWorkflow {
  readonly #memory: Mem0Memory;
  readonly #policy: MemoryDecisionPolicy | undefined;
  readonly #revision: string;
  readonly #limit: number;
  readonly #capturePolicy: { id: string; revision: string } | undefined;
  constructor(memory: Mem0Memory, options: { policy?: MemoryDecisionPolicy; policyRevision?: string; reviewLimit?: number;
    capturePolicyId?: string; capturePolicyRevision?: string } = {}) {
    this.#memory = memory;
    this.#policy = options.policy;
    if (this.#policy !== undefined && typeof this.#policy !== "function") throw new TypeError("policy must be callable");
    this.#revision = requiredText(options.policyRevision ?? "review-only", "policy revision", 512);
    this.#limit = positiveInteger(options.reviewLimit ?? 8, "review limit", 32);
    if ((options.capturePolicyId === undefined) !== (options.capturePolicyRevision === undefined)) {
      throw new TypeError("capture policy id and revision must be configured together");
    }
    this.#capturePolicy = options.capturePolicyId === undefined ? undefined : Object.freeze({
      id: requiredText(options.capturePolicyId, "capture policy id", 512),
      revision: requiredText(options.capturePolicyRevision, "capture policy revision", 512),
    });
  }
  async capture(messages: Parameters<Mem0Memory["extract"]>[0], options: Parameters<Mem0Memory["extract"]>[1]): Promise<MemoryWorkflowResult> {
    const prefix = "workflow:" + digest(requiredText(options.key, "workflow key", 512));
    if (!this.#capturePolicy) {
      if (options.authorization !== undefined) throw new TypeError("capture authorization requires a configured capture policy");
    } else {
      if (options.authorization === undefined) throw new MemoryError("memory_capture_authorization_required");
      const intentOptions = { source: options.source, maxInputChars: this.#memory.maxInputChars,
        ...(options.metadata === undefined ? {} : { metadata: options.metadata }),
        ...(options.expiresAt === undefined ? {} : { expiresAt: options.expiresAt }) };
      const intent = memoryCaptureIntent(messages, intentOptions);
      if (options.authorization.intentDigest !== intent || options.authorization.policyId !== this.#capturePolicy.id
          || options.authorization.policyRevision !== this.#capturePolicy.revision) {
        throw new MemoryError("memory_capture_authorization_mismatch");
      }
    }
    const extraction = await this.#memory.extract(messages, { ...options, key: prefix + ":extract" });
    const resolutions: MemoryOperation[] = [];
    const pendingIds: string[] = [];
    if (extraction.state === "complete") for (const id of extraction.ids) {
      const suffix = digest(this.#revision + "\0" + id);
      const resolveKey = prefix + ":resolve:" + suffix;
      const existing = this.#memory.operation(resolveKey);
      if (existing) {
        resolutions.push(existing);
        if (existing.state !== "complete") pendingIds.push(id);
        continue;
      }
      const signal = options.signal === undefined ? {} : { signal: options.signal };
      const record = await this.#memory.get(id, { includeInactive: true, ...signal });
      if (!record || record.state !== "pending") continue;
      const reviewed = await this.#memory.review({ id, version: record.version }, {
        key: prefix + ":review:" + suffix, limit: this.#limit,
        ...(options.authorization === undefined ? {} : { authorization: options.authorization }), ...signal,
      });
      if (reviewed.state !== "complete" || !reviewed.review || !this.#policy) { pendingIds.push(id); continue; }
      const decision = await this.#policy(record, reviewed.review);
      if (!decision) { pendingIds.push(id); continue; }
      if (decision.reviewKey !== reviewed.review.key || !decision.items.some(ref => ref.id === id && ref.version === record.version)) {
        throw new TypeError("Workflow decision must include the candidate and its review key");
      }
      const resolved = await this.#memory.resolve(decision, { key: resolveKey,
        ...(options.authorization === undefined ? {} : { authorization: options.authorization }), ...signal });
      resolutions.push(resolved);
      if (resolved.state !== "complete") pendingIds.push(id);
    }
    return Object.freeze({ extraction, resolutions: Object.freeze(resolutions), pendingIds: Object.freeze(pendingIds) });
  }
}

File: integrations/mem0/python/src/purra_mem0/workflow.py
"""Restartable extraction/review workflow over the existing operation journal."""

from dataclasses import dataclass
from hashlib import sha256
from typing import Awaitable, Callable

from ._journal import MemoryError
from .memory import (Mem0Memory, MemoryCaptureAuthorization, MemoryOperation, MemoryRecord,
                     MemoryRef, MemoryResolution, MemoryReview, memory_capture_intent)


MemoryDecisionPolicy = Callable[[MemoryRecord, MemoryReview], Awaitable[MemoryResolution | None]]


@dataclass(frozen=True, slots=True)
class MemoryWorkflowResult:
    extraction: MemoryOperation
    resolutions: tuple[MemoryOperation, ...] = ()
    pending_ids: tuple[str, ...] = ()


class MemoryWorkflow:
    """Extract pending records, review them, then ask the host to authorize a resolution.

    No policy means review only. A policy may return None to leave a candidate
    pending. Its revision is part of operation identity; keep it stable on retry.
    Retrieval continues through MemoryContext/assemble_memory_context and only
    uses currently active, authorized records.
    """

    def __init__(self, memory: Mem0Memory, *, policy: MemoryDecisionPolicy | None = None,
                 policy_revision: str = "review-only", review_limit: int = 8,
                 capture_policy_id: str | None = None, capture_policy_revision: str | None = None):
        if not isinstance(policy_revision, str) or not policy_revision.strip() or len(policy_revision) > 512:
            raise ValueError("policy_revision must be non-empty and at most 512 characters")
        if policy is not None and not callable(policy):
            raise TypeError("policy must be callable")
        if type(review_limit) is not int or not 1 <= review_limit <= 32:
            raise ValueError("review_limit must be from 1 to 32")
        if (capture_policy_id is None) != (capture_policy_revision is None):
            raise ValueError("capture policy id and revision must be configured together")
        if capture_policy_id is not None:
            if not isinstance(capture_policy_id, str) or not capture_policy_id.strip() or len(capture_policy_id) > 512:
                raise ValueError("capture_policy_id must be non-empty and at most 512 characters")
            if not isinstance(capture_policy_revision, str) or not capture_policy_revision.strip() or len(capture_policy_revision) > 512:
                raise ValueError("capture_policy_revision must be non-empty and at most 512 characters")
        self.memory, self.policy = memory, policy
        self.policy_revision, self.review_limit = policy_revision, review_limit
        self.capture_policy_id, self.capture_policy_revision = capture_policy_id, capture_policy_revision

    async def capture(self, messages, *, source, key: str, metadata=None, expires_at=None,
                      authorization: MemoryCaptureAuthorization | None = None, signal=None):
        if not isinstance(key, str) or not key.strip() or len(key) > 512:
            raise ValueError("workflow key must be non-empty and at most 512 characters")
        if self.capture_policy_id is None:
            if authorization is not None:
                raise ValueError("capture authorization requires a configured capture policy")
        else:
            if not isinstance(authorization, MemoryCaptureAuthorization):
                raise MemoryError("memory_capture_authorization_required")
            intent = memory_capture_intent(messages, source=source, metadata=metadata, expires_at=expires_at,
                                           max_input_chars=self.memory.max_input_chars)
            if (authorization.intent_digest != intent
                    or authorization.policy_id != self.capture_policy_id
                    or authorization.policy_revision != self.capture_policy_revision):
                raise MemoryError("memory_capture_authorization_mismatch")
        prefix = "workflow:" + sha256(key.encode()).hexdigest()
        # Re-submit the identical extraction input so the journal checks key reuse.
        extraction = await self.memory.extract(messages, source=source, key=prefix + ":extract",
                                               metadata=metadata, expires_at=expires_at,
                                               authorization=authorization, signal=signal)
        if extraction.state != "complete":
            return MemoryWorkflowResult(extraction)
        resolutions, pending = [], []
        for item_id in extraction.ids:
            suffix = sha256((self.policy_revision + "\0" + item_id).encode()).hexdigest()
            resolve_key = prefix + ":resolve:" + suffix
            existing = self.memory.operation(resolve_key)
            if existing is not None:
                resolutions.append(existing)
                if existing.state != "complete":
                    pending.append(item_id)
                continue
            record = await self.memory.get(item_id, include_inactive=True, signal=signal)
            if record is None or record.state != "pending":
                continue  # Never reactivate a revoked or externally resolved record.
            reviewed = await self.memory.review(MemoryRef(record.id, record.version),
                                                key=prefix + ":review:" + suffix,
                                                limit=self.review_limit, signal=signal)
            if reviewed.state != "complete" or reviewed.review is None or self.policy is None:
                pending.append(item_id)
                continue
            decision = await self.policy(record, reviewed.review)
            if decision is None:
                pending.append(item_id)
                continue
            if (not isinstance(decision, MemoryResolution)
                    or decision.review_key != reviewed.review.key
                    or reviewed.review.candidate not in decision.items):
                raise ValueError("workflow decision must include the candidate and its review key")
            # resolve validates all versions, scope, source authority and the reviewed group.
            resolved = await self.memory.resolve(decision, key=resolve_key, signal=signal)
            resolutions.append(resolved)
            if resolved.state != "complete":
                pending.append(item_id)
        return MemoryWorkflowResult(extraction, tuple(resolutions), tuple(pending))


__all__ = ["MemoryWorkflow", "MemoryWorkflowResult", "MemoryDecisionPolicy"]
